Documentation for transient storage (#15409)
ethereum/solidity: docs/contracts/constant-state-variables.rst

Constant and Immutable State Variables
**************************************

> State variables located in storage can be declared as ``constant`` or ``immutable``,

Reviewer: Neither immutables nor constants are located in storage :-).
So maybe just
```suggestion
State variables can be declared as one of ``constant`` or ``immutable`` or ``transient``.
```
Or something like that.

It is also possible to define ``constant`` variables at the file level.

If the state variable is ``transient``, then it cannot be declared as ``constant`` or ``immutable``.

The compiler does not reserve a storage slot for these variables, and every occurrence is

> replaced by the respective value.

Reviewer 2: This remark seems a bit out of place. I'd combine it with the sentence below to make it flow better:

```suggestion
Every occurrence of such a variable in the source is replaced by its underlying value
and the compiler does not reserve a storage slot for it.
It cannot be assigned a slot in transient storage using the ``transient`` keyword either.
```

It is also possible to define ``constant`` variables at the file level.

Every occurrence of such a variable in the source is replaced by its underlying value
and the compiler does not reserve a storage slot for it.

> It cannot be assigned a slot in transient storage using the ``transient`` keyword either.

Reviewer 3: This should perhaps be re-worded, since we forbid constant transient state variables in the first place, so the slot allocation is a moot point anyway.

Reviewer 2: It's basically saying just that. Before it was a remark that you cannot use `transient` with them, which seemed quite random on its own. This says that and indicates why.

Reviewer 3: Fair enough - wasn't saying the statement was wrong in any way, just found it a bit weirdly worded, but I'm fine with it staying as is.
